Develop

braceal opened the pull request.

diangenlin committed:
Update README.md

diangenlin committed:
Write functions to get tsne plots

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Add logger.debug() messages

diangenlin committed:
Modify debug messages

diangenlin committed:
Fix bug related to IsoelectricPoint

diangenlin committed:
Add separate import statement for IsoelectricPoint

diangenlin committed:
Modify tsne save path code

diangenlin committed:
Add functions to get UMAP plots

diangenlin committed:
Update README.md

braceal committed:
fixing logger, moving calculations to metrics.py

diangenlin committed:
Require tsne_path or umap_path is the mode is get_tsne or get_umap, respectively.

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

braceal committed:
fixing logger, moving calculations to metrics.py

braceal committed:
fixing logger, moving calculations to metrics.py

diangenlin committed:
Add back get_embed_avg function

diangenlin committed:
Fix import of biosynseq

diangenlin committed:
Change import biosynseq

diangenlin committed:
Add visualize jupyter notebook files 

Both files are prototypes for visualize.py and metrics.py in biosynseq. The only difference between the two files is that visualize_pairwise2 uses pairwise2 to calculate pairwise alignment scores, while visualze_PairwiseAligner uses PairwiseAligner to do so.

diangenlin committed:
Add visualize jupyter notebook files

diangenlin committed:
Add alignment calculations and plotting code.

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Add missing import statement SeqIO.

diangenlin committed:
Modify if statement in alignment_scores_parallel.

diangenlin committed:
Add missing import statement plt.

diangenlin committed:
Modify README.md

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,375 @@
+    return [len(rec.seq) for rec in seqs]
+
+
+def molecular_weight(protein_seqs: List[SeqRecord]) -> List:
```
What type is each element of the list? The type should be `List[object_type]`. Is it float?

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,375 @@
+    # if alignment_type is not ("global" or "local"):
+    #     raise ValueError(f"Invalid alignment_type: {alignment_type}")
+
+    if (alignment_type != "global") and (alignment_type != "local"):
```
```
if alignment_type not in ("global", "local"):
```

braceal committed:
update to improve efficiency

diangenlin committed:
Add files via upload

Testing using Alex's version 2 functions to calculate alignment scores

diangenlin committed:
fix version 2 of compute alignment scores and parallel functions

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
fix bug in clipping DNA sequences to embedding length

diangenlin committed:
clean up debugger statements and comments

diangenlin committed:
change align plot configs

diangenlin committed:
allow custom title for align plot

diangenlin committed:
Create an_example

diangenlin committed:
Add files via upload

diangenlin committed:
Delete an_example

diangenlin committed:
change plot settings

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Add tsne and umap functions to create subplot graphs for sequence metrics

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,529 @@
+    return seqs
+
+
+def fasta_to_dna_seqs(fasta_path: Path) -> List[SeqRecord]:
```
No need to have this function, simply use the above function.

diangenlin replied:
I had to write this function for the part in the main function of visualize.py to get alignment plots (lines 595-598), where the length of the DNA sequences must be clipped to the length of the embeddings before the DNA sequences can be translated to protein sequences. Clipping is required for the test and validation sequences, but not for generated sequences. This intermediate clipping step thus requires fasta_to_dna_seqs to access the DNA sequences to do the clipping. Do you have suggestions for a better way to go about this?

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,529 @@
+    seqs = list(SeqIO.parse(fasta_path, "fasta"))
+    if translate_to_protein:
+        # translate DNA seqs to protein seqs; stop translation at the first in-frame stop codon
+        seqs = [seq.translate(to_stop=True) for seq in seqs]
```
Call this function here: dna_to_protein_seqs()

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,529 @@
+    return [SeqUtils.molecular_weight(rec.seq, "protein") for rec in protein_seqs]
+
+
+def isoelectric_point(protein_seqs: List[SeqRecord]) -> List:
```
Should be List[float] I believe

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,529 @@
+    return [IsoelectricPoint(seq).pi() for seq in protein_seqs]
+
+
+def compute_alignment_scores(kwargs: dict) -> np.ndarray:
```
Have we moved to using v2 functions? If so, this one (and the one that calls it) can be removed.

braceal commented on biosynseq/visualize.py:
```diff
@@ -1,11 +1,632 @@
+    ValueError
+        If the given tsne_path is not a directory.
+    """
+    df = pd.DataFrame(
```
Is it possible to make a more general plotting function that handles both tsne/umap? This would reduce code duplication.

braceal commented on biosynseq/visualize.py:
```diff
@@ -1,11 +1,632 @@
+    ValueError
+        If the given umap_path is not a directory.
+    """
+    df_dict = {}
```
Same as above: Generalize this to handle both tsne/umap if possible.

braceal commented on biosynseq/visualize.py:
```diff
@@ -1,11 +1,632 @@
+        }
+
+
+def plot_AlignScore_EmbedDist(
```
Call this function plot_embed_dist_vs_align_score

diangenlin committed:
Comment out unused functions and code

diangenlin committed:
Replace fasta_to_dna_seqs with get_seqs_from_fasta

diangenlin committed:
Combine tsne and umap plotting functions

diangenlin committed:
Fix bug in plotting cluster functions

diangenlin committed:
Fix bug in mode tsne/umap of main function

diangenlin committed:
Fix bug in mode align_plot of main function

diangenlin committed:
Remove commented code

diangenlin committed:
Remove commented code

diangenlin committed:
Save README changes

maxzvyagin commented on README.md:
````diff
@@ -78,3 +78,22 @@ python3 generate.py --mode get_fasta --config /homes/lind/MDH-pipeline/mdh_gpt.y
+# Running visualize.py
+## Example 1: Get t-SNE plots
+```
+python3 visualize.py --mode tsne --embed_path /homes/mzvyagin/MDH/perlmutter_data/gpt2_generated_embeddings.npy --fasta_path /homes/mzvyagin/MDH/perlmutter_data/globalstep2850.fasta --cluster_path /homes/lind/MDH-pipeline/visualize/
````
Do we want to have absolute paths on the example? @braceal thoughts?

braceal replied:
It could be good to show a toy example with relative paths to the repository data folder

maxzvyagin commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,403 @@
+
+
+def dna_to_protein_seqs(dna_seqs: List[SeqRecord]) -> List[SeqRecord]:
+    """Translate DNA sequenecs to protein sequences.
```
Lil typo here :)

diangenlin replied:
Nice catch!

maxzvyagin commented on biosynseq/visualize.py:
```diff
@@ -1,11 +1,477 @@
+    np.ndarray
+        Transformed embeddings after running t-SNE.
+    """
+    from cuml.manifold import TSNE
```
Do we want to put an option here to use sklearn if cuml isn't installed? @braceal don't you have it set up like that in mdlearn?

braceal replied:
We could, though best to use RAPIDS. Let's save the update for another day when it's needed.

maxzvyagin commented on biosynseq/visualize.py:
```diff
@@ -1,11 +1,477 @@
+    np.ndarray
+        Transformed embeddings after running UMAP.
+    """
+    from cuml.manifold import UMAP
```
Same as TSNE comment - sklearn option?

maxzvyagin commented on requirements/requirements.txt:
```diff
@@ -3,4 +3,5 @@ scikit-learn
 matplotlib
 plotly
+biopython
 git+https://github.com/ramanathanlab/gene_transformer
```
@braceal So the way we have the gene_transformer repo set up now - will this line in the requirements handle the dependencies for the gene_transformer repo or would those need to get manually installed like in the gene_transformer README?

braceal replied:
They would need to get manually installed I believe.

maxzvyagin commented on requirements/requirements.txt:
```diff
@@ -3,4 +3,5 @@ scikit-learn
 mdlearn
```
Does this install cuml?

braceal replied:
No

maxzvyagin reviewed (commented):
Another suggestion - doesn't necessarily have to be part of this PR. Let's put together a kind of "toy" data example we can put as part of the GitHub? And then the examples can use that versus the lambda paths so folks can actually run it?

diangenlin committed:
Add files via upload

diangenlin committed:
Delete visualize_example.ipynb

diangenlin committed:
Delete notebooks directory

diangenlin committed:
Delete visualize_example_final.ipynb

diangenlin committed:
Create test_example

diangenlin committed:
Add files via upload

diangenlin committed:
Delete test_example

diangenlin committed:
Create README.md

diangenlin committed:
Update README.md

diangenlin committed:
Add files via upload

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,403 @@
+    List[SeqRecord]
+        List of protein sequences.
+    """
+    protein_seqs = [seq.translate(to_stop=True) for seq in dna_seqs]
```
No need for temp variable

diangenlin committed:
Delete visualize_example_final.ipynb

diangenlin committed:
remove temp variable

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Add files via upload

diangenlin committed:
Add umap_random_state argument for UMAP

diangenlin committed:
Add additional UMAP hyperparameters

diangenlin committed:
Add files via upload

diangenlin committed:
Update README.md

diangenlin committed:
Print rec in molecular weight and isoelectric point functions

diangenlin committed:
add str(rec.seq)

diangenlin committed:
Make print statements in molecular_weight optional

diangenlin committed:
modify run_alphafold.py

diangenlin committed:
Add files via upload

diangenlin committed:
Delete visualize_new_model2.ipynb

diangenlin committed:
Update README.md

diangenlin committed:
Update README.md

diangenlin committed:
Update README.md

diangenlin committed:
Update README.md

diangenlin committed:
Add files via upload

diangenlin committed:
Add function to plot metrics histograms

diangenlin committed:
Add function to plot metrics histograms

diangenlin committed:
Modify doctrings

diangenlin committed:
Change default value in plot_metrics_hist

diangenlin committed:
Fix rounding bug

diangenlin committed:
Fix bug in rounding

diangenlin committed:
Fix bug in calculating number of rows in subplot

diangenlin committed:
Update README.md

diangenlin committed:
Fix bug in calculating nrows for subplots

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Clean up unused code

diangenlin committed:
Add files via upload

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,421 @@
+    List
+        Molecular weight of each protein.
+    """
+    # return [SeqUtils.molecular_weight(rec.seq, "protein") for rec in protein_seqs]
```
The original implementation can be used again. The debug version can be removed

braceal commented on biosynseq/metrics.py:
```diff
@@ -0,0 +1,421 @@
+    List
+        Isoelectric point of each protein.
+    """
+    # return [IsoelectricPoint(seq).pi() for seq in protein_seqs]
```
Original version should be fine

braceal committed:
mypy

braceal committed:
rm debug logic

braceal committed:
update type

diangenlin committed:
Add function to compute mean, min, max alignment scores

braceal committed:
account for case when num workers > num seqs

braceal committed:
generalize logic

braceal committed:
remove duplicate code

diangenlin committed:
Add functions to compute mean/min/max alignment scores and plot histogram of these scores

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Add docstrings

braceal committed:
refactor

braceal committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
fix typo

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

braceal committed:
name typo

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Change all saving paths to save_dir

diangenlin committed:
Change plot_cluster and plot_cluster_subplots to be tsne/umap agnostic

diangenlin committed:
Add functions to main() function

braceal committed:
rm mdlearn, clean func

braceal committed:
merge conflicy

diangenlin committed:
Change save_dir's default value to Path()

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Create test

Create the new folder Dateset_Examples to put Jupyter Notebooks containing analysis of datasets coming from other DL models.

diangenlin committed:
Delete test

diangenlin committed:
Create test

Create a new folder named Dataset_Examples to upload Jupyter Notebook analyses of sequence datasets coming from different DL models.

braceal committed:
diverse beta sampling, phylogenetic tree from latent, visualize generated with train seqs

braceal committed:
:qMerge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Change all save_path settings to be optional

diangenlin committed:
Merge branch 'develop' of https://github.com/ramanathanlab/SyntheticSequenceEvaluation into develop

diangenlin committed:
Fix save_path bug in plot_cluster

diangenlin committed:
modify suffix for get_clusters

diangenlin committed:
Fix bug in suffix

diangenlin committed:
Fix bug in suffix

diangenlin committed:
fix suffix issue

diangenlin committed:
Delete visualize_example.ipynb

diangenlin committed:
Add files via upload

diangenlin committed:
visual analysis for different datasets/models

Datasets and models: MDH-GPT-2, MDH-reformer, and spike protein-reformer

diangenlin committed:
Delete visualize_covid_spike_reformer.ipynb

diangenlin committed:
Delete visualize_mdh_gpt2_likely-wood-857_fixed_bug.ipynb

diangenlin committed:
Delete visualize_mdh_reformer.ipynb

diangenlin committed:
Delete visualize_new_model.ipynb

diangenlin committed:
Update README.md

Modify the example command line code for plotting various plots enabled by visualize.py.

diangenlin committed:
remove extra print statement for plotting

braceal committed:
updating API to latest version of gene_transformer

braceal committed:
generating valid sequences notebook

diangenlin committed:
Delete test

maxzvyagin commented on biosynseq/generate.py:
```diff
@@ -40,14 +40,12 @@ def generate_fasta(
-    model.cuda()
-    # generate non-redundant sequences
+    # Obtain model
+    model = model_strategy.get_model(DNATransformer).cuda()
```
Do we want to make the cuda call an option here? I think it's fine hard coded but something to consider :)

maxzvyagin commented on biosynseq/generate.py:
```diff
@@ -155,18 +159,17 @@ def main() -> None:
-        model_strategy = LoadPTCheckpointStrategy(config, args.pt_path)
-    elif args.inference_model_load == "deepspeed":
-        model_strategy = LoadDeepSpeedStrategy(config)
+    if args.load_pt_checkpoint is not None:
```
Do we want to add a check if both load_pt_checkpoint *and* load_ds_checkpoint are set? We can do that check and if both are set simply add a warning call that prints out saying we use pt_checkpoint if both are set.

maxzvyagin reviewed (approved):
Looks good! I added a couple comments but I'd be good merging as is now.

diangenlin committed:
Add files via upload

diangenlin committed:
Update README.md

diangenlin committed:
Add files via upload

Selected MDH generated sequences (GPT-2, likely-wood-857).

diangenlin committed:
Add files via upload

Upload file for global alignment between likely-wood-857 generated/train sequences and 2pwz sequences

diangenlin committed:
visualize mdh gpt2 likely-wood-857 with sklearn

diangenlin committed:
visualize mdh gpt2 likely-wood-857 full

diangenlin committed:
Delete visualize_align_likely-wood-857_vs_2pwz.ipynb

diangenlin committed:
visualize mdh gpt2 likely-wood-857 vs 2pwz

diangenlin committed:
Delete visualize_mdh_gpt2_likely-wood-857_final2.ipynb

diangenlin committed:
Delete visualize_mdh_gpt2_likely-wood-857.ipynb

diangenlin committed:
Delete visualize_example.ipynb

diangenlin committed:
visualize example

diangenlin committed:
Create test_msa_aln

diangenlin committed:
MSA alignments of likely-wood-857 sequences

diangenlin committed:
Create tree_test

diangenlin committed:
tree and xml files of likely-wood-857 sequences

diangenlin committed:
Delete test_msa_aln

diangenlin committed:
Delete tree_test

diangenlin committed:
phylogenetic analysis of likely-wood-857 sequences

diangenlin committed:
changes to visualize.py

braceal committed:
Merge branch 'main' into develop

braceal committed:
line counts
